File: app/src/main/java/org/pacien/tincapp/commands/Tinc.java
package org.pacien.tincapp.commands;

import android.content.Context;

import com.annimon.stream.Stream;

import org.pacien.tincapp.context.AppPaths;

import java.io.IOException;
import java.util.List;

/**
 * @author pacien
 */
final public class Tinc {

	private Tinc() {
		// static class
	}

	static private Command newCommand(Context ctx, String netName) {
		return new Command(AppPaths.tinc(ctx).getAbsolutePath())
				.withOption("config", AppPaths.confDir(ctx, netName).getAbsolutePath())
				.withOption("pidfile", AppPaths.pidFile(ctx, netName).getAbsolutePath());
	}

	// independently runnable commands

	static public List<String> network(Context ctx) throws IOException {
		return Executor.call(new Command(AppPaths.tinc(ctx).getAbsolutePath())
				.withOption("config", AppPaths.confDir(ctx).getAbsolutePath())
				.withArguments("network"));
	}

	static public List<String> fsck(Context ctx, String netName, Boolean fix) throws IOException {
		Command cmd = newCommand(ctx, netName).withArguments("fsck");
		if (fix) cmd = cmd.withOption("force");
		return Executor.call(cmd);
	}

	// commands requiring a running tinc daemon

	static public void stop(Context ctx, String netName) throws IOException {
		Executor.call(newCommand(ctx, netName).withArguments("stop"));
	}

	static public List<String> dumpNodes(Context ctx, String netName, Boolean reachable) throws IOException {
		Command cmd = reachable
				? newCommand(ctx, netName).withArguments("dump", "reachable", "nodes")
				: newCommand(ctx, netName).withArguments("dump", "nodes");

		return Executor.call(cmd);
	}

	static public String info(Context ctx, String netName, String node) throws IOException {
		List<String> output = Executor.call(newCommand(ctx, netName).withArguments("info", node));
		return Stream.of(output).reduce((l, r) -> l + '\n' + r).get();
	}

}

File: app/src/main/java/org/pacien/tincapp/commands/Tinc.kt
/*
 * Tinc App, an Android binding and user interface for the tinc mesh VPN daemon
 * Copyright (C) 2017-2018 Pacien TRAN-GIRARD
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.pacien.tincapp.commands

import java8.util.concurrent.CompletableFuture
import org.pacien.tincapp.context.AppPaths

/**
 * @author pacien
 */
object Tinc {
  private fun newCommand(netName: String): Command =
    Command(AppPaths.tinc().absolutePath)
      .withOption("config", AppPaths.confDir(netName).absolutePath)
      .withOption("pidfile", AppPaths.pidFile(netName).absolutePath)

  fun stop(netName: String): CompletableFuture<Unit> =
    Executor.call(newCommand(netName).withArguments("stop"))
      .thenApply { }

  fun pid(netName: String): CompletableFuture<Int> =
    Executor.call(newCommand(netName).withArguments("pid"))
      .thenApply { Integer.parseInt(it.first()) }

  fun dumpNodes(netName: String, reachable: Boolean = false): CompletableFuture<List<String>> =
    Executor.call(
      if (reachable) newCommand(netName).withArguments("dump", "reachable", "nodes")
      else newCommand(netName).withArguments("dump", "nodes"))

  fun info(netName: String, node: String): CompletableFuture<String> =
    Executor.call(newCommand(netName).withArguments("info", node))
      .thenApply<String> { it.joinToString("\n") }

  fun init(netName: String, nodeName: String): CompletableFuture<String> =
    if (netName.isBlank())
      CompletableFuture.failedFuture(IllegalArgumentException("Network name cannot be blank."))
    else
      Executor.call(Command(AppPaths.tinc().absolutePath)
        .withOption("config", AppPaths.confDir(netName).absolutePath)
        .withArguments("init", nodeName))
        .thenApply<String> { it.joinToString("\n") }

  fun join(netName: String, invitationUrl: String): CompletableFuture<String> =
    if (netName.isBlank())
      CompletableFuture.failedFuture(IllegalArgumentException("Network name cannot be blank."))
    else
      Executor.call(Command(AppPaths.tinc().absolutePath)
        .withOption("config", AppPaths.confDir(netName).absolutePath)
        .withArguments("join", invitationUrl))
        .thenApply<String> { it.joinToString("\n") }

  fun log(netName: String, level: Int? = null): Process =
    Executor.run(newCommand(netName)
      .withArguments("log")
      .apply { if (level != null) withArguments(level.toString()) })
}
